python/qemu/ot/util/misc.py

```diff
@@ -193,10 +196,83 @@ def to_bool(value, permissive=True, prohibit_int=False):
         return False
     raise ValueError(f"Invalid boolean value: '{value}")
 
+
 def alphanum_key(text: str) -> list[Union[int, str]]:
     """Alphanumerical sorting key.
 
        :param text: the text to generate a sorting key from
        :return: a list of alternating str and integer values
     """
     return [int(t) if t.isdigit() else t for t in re.split(r'(\d+)', text)]
+
+
+def redent(text: str, spc: int = 0, strip_end: bool = False) -> str:
+    """Utility function to re-indent code string.
+
+        :param text: the text to re-indent
+        :param spc: the number of leading empty space chars to prefix lines
+        :param strip_end: whether to strip trailing whitespace and newline
+    """
+    text = dedent(text.lstrip('\n'))
+    text = indent(text, ' ' * spc)
+    if strip_end:
+        text = text.rstrip(' ').rstrip('\n')
+    return text
+
+
+def retrieve_git_version(path: str, max_tag_dist: int = 100) \
+        -> Optional[str]:
+    """Return a Git identifier whenever possible.
+
+        :param path: the configuration file or directory to track the repository
+                     version identifier. Note that the returned Git identifier
+                     is not the commit version of this file / directory, but the
+                     one of the repo top-level directory.
+        :param max_tag_dist: maximum distance (in commit number) to the closest
+                             tag. If distance is greater, only emit the commit
+                             identifier
+        :return: Git version of the top-level directory
+    """
+    cfgdir: Optional[str] = None
+    path = abspath(path)
+    if isfile(path):
+        cfgdir = dirname(path)
+    elif isdir(path):
+        cfgdir = path
+    else:
+        return None
+    while cfgdir and isdir(cfgdir):
+        gitdir = joinpath(cfgdir, '.git')
+        if exists(gitdir):  # either a dir or a file for worktree
+            break
+        cfgdir = dirname(cfgdir)
+    else:
+        return None
+    assert cfgdir is not None
+    try:
+        descstr = check_output(['git', 'describe', '--long', '--dirty'],
+                               text=True, cwd=cfgdir).strip()
+        gmo = re.match(r'^(?P<tag>.*)-(?P<dist>\d+)-g(?P<commit>[0-9a-f]+)'
+                       r'(?:-(?P<dirty>dirty))?$', descstr)
+        if not gmo:
+            raise ValueError('Unknown Git describe format')
```

Comment:
Maybe this could log the output?

```suggestion
            raise ValueError(f'Unknown Git describe format: {descstr}')
```

Reply:
I'm not sure how to handle this since `ValueError` is caught and discarded, _i.e._ the message is more a comment that something that gets reported. There is no logger here, and printing out to stderr is quite intrusive.

Reply:
I see, would it not be better to do anyway in case code is added in the future that relies on this function but does not discard the error?

Reply:
It does not hurt adding it, but using it would require to modify the function itself since it is the function that discards the ValueError (line 266/267)

Reply:
Ah, sorry, I missed the context that this was in a `try: ... except: ...` when I replied. Then yes, I guess it is fine to leave it as is.